In [1]:
# import required libraries and read the data
import warnings
warnings.filterwarnings('ignore')
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
import math

df=pd.read_csv('Credit.csv')
df.head()

Income  Limit  Rating  Cards  Age  Education OwnHome Student Married  \
0   14.891   3606     283      2   34         11      No      No     Yes   
1  106.025   6645     483      3   82         15     Yes     Yes     Yes   
2  104.593   7075     514      4   71         11      No      No      No   
3  148.924   9504     681      3   36         11     Yes      No      No   
4   55.882   4897     357      2   68         16      No      No     Yes   

  Region  Balance  
0  South      333  
1   West      903  
2   West      580  
3   West      964  
4  South      331

In [2]:
df.shape

(400, 11)

In [3]:
# check for missing values
df.isnull().sum()

Income       0
Limit        0
Rating       0
Cards        0
Age          0
Education    0
OwnHome      0
Student      0
Married      0
Region       0
Balance      0
dtype: int64

In [5]:
# exploratory data analysis
df.describe()

Income         Limit      Rating       Cards         Age  \
count  400.000000    400.000000  400.000000  400.000000  400.000000   
mean    45.218885   4735.600000  354.940000    2.957500   55.667500   
std     35.244273   2308.198848  154.724143    1.371275   17.249807   
min     10.354000    855.000000   93.000000    1.000000   23.000000   
25%     21.007250   3088.000000  247.250000    2.000000   41.750000   
50%     33.115500   4622.500000  344.000000    3.000000   56.000000   
75%     57.470750   5872.750000  437.250000    4.000000   70.000000   
max    186.634000  13913.000000  982.000000    9.000000   98.000000   

        Education      Balance  
count  400.000000   400.000000  
mean    13.450000   520.015000  
std      3.125207   459.758877  
min      5.000000     0.000000  
25%     11.000000    68.750000  
50%     14.000000   459.500000  
75%     16.000000   863.000000  
max     20.000000  1999.000000

In [6]:
# get binary variables of categorical columns
df=pd.get_dummies(df)
df.head()

Income  Limit  Rating  Cards  Age  Education  Balance  OwnHome_No  \
0   14.891   3606     283      2   34         11      333           1   
1  106.025   6645     483      3   82         15      903           0   
2  104.593   7075     514      4   71         11      580           1   
3  148.924   9504     681      3   36         11      964           0   
4   55.882   4897     357      2   68         16      331           1   

   OwnHome_Yes  Student_No  Student_Yes  Married_No  Married_Yes  Region_East  \
0            0           1            0           0            1            0   
1            1           0            1           0            1            0   
2            0           1            0           1            0            0   
3            1           1            0           1            0            0   
4            0           1            0           0            1            0   

   Region_South  Region_West  
0             1            0  
1             0            1  
2             0            1  
3             0            1  
4             1            0

In [7]:
# create correlation matrix
corr = df.corr()
corr.style.background_gradient()

In [8]:
# set dependent and independent variables
X = df[['Rating', 'Student_Yes']].copy()
Y = df['Balance'].copy()

In [9]:
# split the data
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.3, random_state = 0)

In [10]:
# run OLS Regression model on Train set
X_train = sm.add_constant(X_train) 
est = sm.OLS(Y_train,X_train).fit()
predictions = est.predict() 
print(est.summary())
print("\nAverage error: {:.2f}.".format(math.sqrt(est.mse_resid)))

                            OLS Regression Results                            
Dep. Variable:                Balance   R-squared:                       0.814
Model:                            OLS   Adj. R-squared:                  0.813
Method:                 Least Squares   F-statistic:                     606.1
Date:                Fri, 05 Jan 2024   Prob (F-statistic):          6.80e-102
Time:                        05:20:38   Log-Likelihood:                -1881.6
No. Observations:                 280   AIC:                             3769.
Df Residuals:                     277   BIC:                             3780.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const        -434.0290     29.730    -14.599      

In [11]:
# use K-Fold cross validation 
kf =KFold(n_splits=5, shuffle=True, random_state=0)
lr = LinearRegression()
CVscores= cross_val_score(lr, X_train, Y_train, scoring='r2', cv=kf)
print('Cross validation R-Squared ', CVscores)
print('Cross validation R-Squared mean:', CVscores.mean())

Cross validation R-Squared  [0.81296453 0.79167939 0.79268713 0.78159709 0.85738448]
Cross validation R-Squared mean: 0.8072625241253961


K-Fold cross validation is used to evaluate machine learning models by providing for more accurate measure of performance. It splits the dataset into equally sized subsets (folds) where one fold acts as test set and the others as train sets. This process repeats until all folds acted as a test fold. Scores are retained after each iteration and then averaged to asses the overall performance of the model. 

The average R^2 from the cross validation is 0.807 which is 0.07 less than the score from the model on Train set. This indicates that the model has an overfitting problem, but because the difference is very small it could be ignored

In [13]:
# check for multicollinearity, VIF for the features must be less than 2
variables = est.model.exog
vif = pd.DataFrame() 
vif["VIF Factor"] = [variance_inflation_factor(variables, i) for i in range(variables.shape[1])]
vif["features"] = X_train.columns
print('VIF: {}'.format(vif))

VIF:    VIF Factor     features
0    6.085428        const
1    1.005715       Rating
2    1.005715  Student_Yes


In [14]:
# run the regression on the Test set
X_test = sm.add_constant(X_test)
est = sm.OLS(Y_test,X_test).fit()
predictions = est.predict()
print(est.summary())
print("\nAverage error: {:.2f}.".format(math.sqrt(est.mse_resid)))

                            OLS Regression Results                            
Dep. Variable:                Balance   R-squared:                       0.814
Model:                            OLS   Adj. R-squared:                  0.811
Method:                 Least Squares   F-statistic:                     256.7
Date:                Fri, 05 Jan 2024   Prob (F-statistic):           1.62e-43
Time:                        05:24:57   Log-Likelihood:                -800.76
No. Observations:                 120   AIC:                             1608.
Df Residuals:                     117   BIC:                             1616.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const        -434.4279     48.857     -8.892      

In [15]:
# multicollinearity check
variables = est.model.exog
vif = pd.DataFrame() 
vif["VIF Factor"] = [variance_inflation_factor(variables, i) for i in range(variables.shape[1])]
vif["features"] = X_test.columns
print('VIF: {}'.format(vif))

VIF:    VIF Factor     features
0    7.628032        const
1    1.034316       Rating
2    1.034316  Student_Yes


**Interpret the regression results:**

**R^2** - 0.814 meaning that 81.4% of change in price can be explained by Rating and Student_Yes

**Adj. R^2** - Adj. R^2 is lower than R^2 indicating some variables are not contributing to the model's R^2 properly, but they are very close so it should not be a problem

**Prob (F-statistic)** - Prob (F-statistic) is 1.62x10^-43 which is less than 0.05 showing the model is linear and significant

**coefficients** - for 1 unit increase in Rating the Balance will increase by 2.61, and relative to the Student_No, Student_Yes have on average higher Balance by 354.49 

**p-values** - p-values for both independent variables are less than 0.05 meaning they are both statistically significant

**Average Error** - The Average Error is 193.78 meaning that on average the model is 193.78 off

The average errors are different between the Test and Train sets, and the R^2 for the both models is the same but the Adjusted R^2 for the Train set is 0.02 higher than for the Test set indicating very slight overfitting problem

The model is generalizable to the unseen data because even though there is a difference between the Test set and Train set in Average Error and Adj. R^2 resulting in an overfitting problem, but the difference is very small and the R^2 is same for the two sets, therefore, the model should generalize well to any new data.